Allow HOMEBREW_CURLRC to specify a path for curl `--config`

clint-stripe opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

We're using a custom `HOMEBREW_ARTIFACT_DOMAIN` that requires some additional flags in `.curlrc`. We would like to set `HOMEBREW_CURLRC` and the value for `CURL_HOME` in `/etc/homebrew/brew.env` to support this without relying on the user's `.curlrc`.

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -136,6 +136,10 @@ module EnvConfig
         boolean:     true,
       },
+      HOMEBREW_CURL_HOME:                        {
+        description: "If set alongside `HOMEBREW_CURLRC`, `curl`(1) will be invoked with `CURL_HOME` " \
```
What is `CURL_HOME` used for beyond finding `.curlrc`?

clint-stripe replied:
I believe that's the only thing -- ideally we could pass a path to `.curlrc` rather than its parent directory, but that doesn't appear possible: https://everything.curl.dev/cmdline/configfile#default-config-file

MikeMcQuaid replied:
We could perhaps handle that in `HOMEBREW_CURLRC` logic itself so that we set `CURL_HOME` internally to look in the full path specified.

clint-stripe replied:
Okay, reading the docs _once_ more I see there is actually a `--config` option that could be used!

Is your proposal to change `HOMEBREW_CURLRC` from a boolean to a file path? (In that case I suppose if it's set but _not_ a valid path, we could use the old behavior for compatibility? 😬)

clint-stripe replied:
I gave that a try, and tested in my environment that needs this. I don't mind that approach at all, actually! Let me know what you think.

clint-stripe commented on bin/brew:
```diff
@@ -205,6 +205,13 @@ ENV_VAR_NAMES=(
+# When using .curlrc is permitted, also allow setting CURL_HOME
+if [[ -n "${HOMEBREW_CURLRC-}" && -n "${HOMEBREW_CURL_HOME-}" ]]
+then
+  FILTERED_ENV+=("CURL_HOME=${HOMEBREW_CURL_HOME}")
```
That's fair! I'll give that a try.

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -132,9 +132,9 @@ module EnvConfig
-        description: "If set, do not pass `--disable` when invoking `curl`(1), which disables the " \
-                     "use of `curlrc`.",
-        boolean:     true,
+        description: "If set to a path, pass it with `--config` when invoking `curl`(1). " \
+                     "If set but _not_ a valid path, do not pass `--disable`, which disables the " \
+                     "use of `.curlrc`.",
```
This is great. Wondering if this is the best approach or if we want to do something like "if set to `1` or `true`"? @Homebrew/brew thoughts?

clint-stripe replied:
That's a very fair question! The current handling of `boolean: true` is `#present?`, but it might be nice to enforce something stricter in this particular case.

Bo98 replied:
> This is great. Wondering if this is the best approach or if we want to do something like "if set to `1` or `true`"? @Homebrew/brew thoughts?

Starting with `/` is probably sufficient to invoke the path-based logic.

There's potentially also an argument that we could just let `CURL_HOME` passthrough if `HOMEBREW_CURLRC` is set rather than the double-purpose, but it maybe doesn't matter much.

MikeMcQuaid replied:
> Starting with `/` is probably sufficient to invoke the path-based logic, given it is unlikely someone does that in order to use it as a boolean and we won't ever support relative paths.

Agreed, that works for me!

MikeMcQuaid commented on Library/Homebrew/test/utils/curl_spec.rb:
```diff
@@ -313,11 +313,27 @@
       expect(curl_args(*args).first).not_to eq("--disable")
     end
 
+    it "doesn't return `--config` when HOMEBREW_CURLRC is unset" do
+      expect(curl_args(*args)).not_to include(a_string_starting_with("--config="))
+    end
+
+    it "returns `--config` when HOMEBREW_CURLRC is a valid path" do
+      Tempfile.create do |tmpfile|
+        path = tmpfile.path
+        ENV["HOMEBREW_CURLRC"] = path
+        # We still expect --disable
+        expect(curl_args(*args).first).to eq("--disable")
+        expect(curl_args(*args).join(" ")).to include("--config #{path}")
+      end
+    ensure
+      ENV["HOMEBREW_CURLRC"] = nil
+    end
```
😍

MikeMcQuaid reviewed (commented):
Great work here @clint-stripe! Excuse all the back-and-forth: just the first time we've had a config change like this so want to make sure we get it right first time 😎

MikeMcQuaid commented on Library/Homebrew/env_config.rb:
```diff
@@ -132,9 +132,9 @@ module EnvConfig
-        description: "If set, do not pass `--disable` when invoking `curl`(1), which disables the " \
-                     "use of `curlrc`.",
-        boolean:     true,
+        description: "If set to a path, pass it with `--config` when invoking `curl`(1). " \
+                     "If set but _not_ a valid path, do not pass `--disable`, which disables the " \
+                     "use of `.curlrc`.",
```
```suggestion
        description: "If set to a full path (i.e. beginning with `/`, pass it with `--config` when invoking `curl`(1). " \
                     "If set but _not_ a valid path, do not pass `--disable`, which disables the " \
                     "use of `.curlrc`.",
```

note this will fail `brew style`: you'll need to fix up the strings locally. Once that's done, you've run `brew generate-man-completions` and (ideally if your git-fu is sufficient) squashed it down to a single commit and CI is 🟢: this is ready to merge! Sorry for all the back and forth.

clint-stripe commented:
Awesome, thank you so much for the quick reviews! (Also really appreciate `brew tests`, `brew style`, etc 🙂)

I made that last change you suggested, fixed up the line lengths, and squashed.

carlocab commented on Library/Homebrew/env_config.rb:
```diff
@@ -132,9 +132,10 @@ module EnvConfig
-        description: "If set, do not pass `--disable` when invoking `curl`(1), which disables the " \
-                     "use of `curlrc`.",
-        boolean:     true,
+        description: "If set to a full path (i.e. beginning with `/`), pass it with `--config` when invoking " \
```
```suggestion
        description: "If set to an absolute path (i.e. beginning with `/`), pass it with `--config` when invoking " \
```
Maybe? "absolute" strikes me as more standard terminology, but not a strong opinion.

MikeMcQuaid reviewed (approved):
Thanks so much for your first contribution (hopefully of many)! Without people like you submitting PRs we couldn't run this project. You rock, @clint-stripe!
